src/EventSubscriber/ActiveSessionListener.php

```diff
@@ -0,0 +1,89 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+declare(strict_types = 1);
+
+namespace Surfnet\Tiqr\EventSubscriber;
+
+use Psr\Log\LoggerInterface;
+use Surfnet\Tiqr\Attribute\RequiresActiveSession;
+use Surfnet\Tiqr\Service\SessionCorrelationIdService;
+use Surfnet\Tiqr\WithContextLogger;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\ControllerArgumentsEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+use function is_array;
+
+/**
+ * Listen to all incoming requests and log the session state information.
+ *
+ * The listener also looks for the #[RequiresActiveSession] attribute.
+ * When a route is marked as to have a required active session this listener will deny access when there is none.
+ */
+final readonly class ActiveSessionListener implements EventSubscriberInterface
+{
+    public function __construct(
+        private LoggerInterface             $logger,
+        private SessionCorrelationIdService $sessionCorrelationIdService
+    ) {
+    }
+
+    public function onKernelControllerArguments(ControllerArgumentsEvent $event): void
+    {
+        $logger = WithContextLogger::from($this->logger, [
+            'correlationId' => $this->sessionCorrelationIdService->generateCorrelationId() ?? '',
+            'route' => $event->getRequest()->getRequestUri(),
+        ]);
+
+        $sessionCookieId = $event->getRequest()->cookies->get('PHPSESSID');
+        if ($sessionCookieId !== null) {
+            $logger->info('User made a request with a session cookie.');
+        }
+
+        // If the session cookie matches the session id we can stop the execution.
+        $sessionId = $event->getRequest()->getSession()->getId();
+        if ($sessionCookieId === $sessionId) {
+            $logger->info('User session matches the session cookie.');
+            return;
+        }
+
+        $logger->error('The session cookie does not match the session id.');
+
+        // If the route does not require an active session we can stop the execution.
+        if (!is_array($event->getAttributes()[RequiresActiveSession::class] ?? null)) {
+            return;
+        }
+
+        // The route requires an active session, no authentication was found.
+        // We'll log an error that differentiates between whether a session cookie was set, or not.
+        if ($sessionCookieId !== null) {
+            $logger->error('Route requires active session. Active session wasn\'t found.');
+            throw new AccessDeniedException();
+        }
```

Comment:
When the session cookie id is not null, that implies there is a session present right? But you log an error? Might be I'm overlooking something essential here. But this is kind of confusing to me. It feels to me this should be the place where you can pass the validation.

Reply:
If there is a session cookie _and_ there is a session id _and_ both of them are one and the same, the function execution has been stopped before and this code isn't reached. You can find the code responsible for this on line 61.

```
        // If the session cookie matches the session id we can stop the execution.
        $sessionId = $event->getRequest()->getSession()->getId();
        if ($sessionCookieId === $sessionId) {
            $logger->info('User session matches the session cookie.');
            return;
        }

```